### PR TITLE
Add optimization section to readme files

### PR DESCRIPTION
**Pull request by LeonidBeynenson**

### CODE REVIEW

File: models/image_classification/model_templates/custom-classification/README.md [hunk 1]
```diff
@@ -190,3 +190,29 @@ python export.py \
 
 This produces model `model.xml` and weights `model.bin` in single-precision floating-point format
 (FP32). The obtained model expects **normalized image** in planar BGR format.
+
+
+### 8. Optimization
+
+The models may be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
```
Comment (snosov1) [inline models/image_classification/model_templates/custom-classification/README.md:197]:
```suggestion
The models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
```

File: models/image_classification/model_templates/custom-classification/README.md [hunk 2]
```diff
@@ -190,3 +190,29 @@ python export.py \
...
+
+At the moment the only one compression methods is supported for image classification models: int8 quantization.
```
Comment (snosov1) [inline models/image_classification/model_templates/custom-classification/README.md:199]:
```suggestion
At the moment, only one compression methods is supported for image classification models: int8 quantization.
```

File: models/image_classification/model_templates/custom-classification/README.md [hunk 3]
```diff
@@ -190,3 +190,29 @@ python export.py \
...
+   --load-weights ${SNAPSHOT} \
+   --train-ann-files '' \
+   --train-data-roots ${TRAIN_DATA_ROOT} \
+   --val-ann-files '' \
+   --val-data-roots ${VAL_DATA_ROOT} \
+   --save-checkpoints-to outputs \
+   --nncf-quantization
+```
+
+The compressed model may be evaluated and exported to the OpenVINO™ format by the same commands as non-compressed model,
```
Comment (snosov1) [inline models/image_classification/model_templates/custom-classification/README.md:217]:
```suggestion
The compressed model can be evaluated and exported to the OpenVINO™ format by the same commands as non-compressed model,
```

**Comment by MaximProshin** [comment]
@LeonidBeynenson , I propose to also add @AlexanderDokuchaev  and @AlexKoff88 to reviewers.
@aleksei-kashapov FYI when you're back from the vacation.

File: models/object_detection/model_templates/custom-object-detection/readme.md [hunk 4]
```diff
@@ -140,3 +140,39 @@ python eval.py \
...
    --classes ${CLASSES}
 ```
+
+### 11. Optimization
+
+The trained models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, the following compression methods are supported:
+* int8 quantization
+* sparsity
```
Comment (AlexanderDokuchaev) [inline models/object_detection/model_templates/custom-object-detection/readme.md:150]: Sparsity and filter pruning was not added. 
I think that will be better to add link to compression_config.json.

Something like
```
List of available compression methods you can find in compression_config.json.
```

Comment (AlexanderDokuchaev) [reply models/object_detection/model_templates/custom-object-detection/readme.md:150]: PS. sparsity config exists only for face-detection-0200

File: models/object_detection/model_templates/custom-object-detection/readme.md [hunk 5]
```diff
@@ -140,3 +140,39 @@ python eval.py \
...
+
+Note that in the command above the parameter `--nncf-quantization` is used to make int8 quantization of the model.
+You can use the following command line flags to turn on compression methods:
+* `--nncf-quantization` -- int8 compression
+* `--nncf-sparsity` -- sparsity
+* `--nncf-pruning` -- filter pruning
+
+The combinations of several compression methods are also possible, but were not tested thoroughly.
+
+The compressed model can be evaluated and exported to the OpenVINO™ format by the same commands as non-compressed model,
```
Comment (AlexanderDokuchaev) [inline models/object_detection/model_templates/custom-object-detection/readme.md:177]: May be add move optimization step befor eval and export?

File: models/object_detection/model_templates/custom-object-detection/readme.md (hunk 4 above)
Comment (AlexKoff88) [reply models/object_detection/model_templates/custom-object-detection/readme.md:150]: I suggest removing this config since it is not validated and planned to be used.

File: models/object_detection/model_templates/custom-object-detection/readme.md (hunk 5 above)
Comment (AlexKoff88) [reply models/object_detection/model_templates/custom-object-detection/readme.md:177]: I think the current flow with Train->Eval->Export->Optimize is ok.

**Comment by MaximProshin** [comment]
General comment: I think we should specify a particular compression algorithm in those places where it's only supported, i.e. INT8 Quantization only where it's supported and Filter Pruning where it's only supported.

File: models/object_detection/model_templates/custom-object-detection/readme.md [hunk 6]
```diff
@@ -140,3 +140,39 @@ python eval.py \
    --save-metrics-to ${WORK_DIR}/metrics.yaml \
    --classes ${CLASSES}
 ```
+
+### 11. Optimization
+
+The trained models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, the following compression methods are supported:
+* int8 quantization
```
Comment (AlexanderDokuchaev) [inline models/object_detection/model_templates/custom-object-detection/readme.md:149]: On current moment added only quantization configs

File: models/object_detection/model_templates/face-detection/readme.md [hunk 7]
```diff
@@ -132,3 +132,40 @@ Try both following variants and select the best one:
...
 As soon as training is completed, it is worth to re-evaluate trained model on test set (see Step 4.b).
+
+
+### 5. Optimize
+
+The models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, the following compression methods are supported:
+* int8 quantization
+* sparsity
```
Comment (AlexanderDokuchaev) [inline models/object_detection/model_templates/face-detection/readme.md:143]: Pruning and sparsity added only for some

File: models/object_detection/model_templates/person-detection/readme.md [hunk 8]
```diff
@@ -128,3 +128,40 @@ Try both following variants and select the best one:
...
 
 As soon as training is completed, it is worth to re-evaluate trained model on test set (see Step 4.b).
+
+
+### 5. Optimize
+
+The models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, the following compression methods are supported:
+* int8 quantization
```
Comment (AlexanderDokuchaev) [inline models/object_detection/model_templates/person-detection/readme.md:138]: On current moment added only quantization configs

File: models/object_detection/model_templates/person-vehicle-bike-detection/readme.md [hunk 9]
```diff
@@ -129,3 +129,40 @@ Try both following variants and select the best one:
...
 
 As soon as training is completed, it is worth to re-evaluate trained model on test set (see Step 4.b).
+
+
+### 5. Optimize
+
+The models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, the following compression methods are supported:
+* int8 quantization
```
Comment (AlexanderDokuchaev) [inline models/object_detection/model_templates/person-vehicle-bike-detection/readme.md:139]: On current moment added only quantization configs

File: models/object_detection/model_templates/vehicle-detection/readme.md [hunk 10]
```diff
@@ -128,3 +128,40 @@ Try both following variants and select the best one:
...
 
 As soon as training is completed, it is worth to re-evaluate trained model on test set (see Step 4.b).
+
+
+### 5. Optimize
+
+The models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, the following compression methods are supported:
+* int8 quantization
```
Comment (AlexanderDokuchaev) [inline models/object_detection/model_templates/vehicle-detection/readme.md:138]: On current moment added only quantization configs

**Review by AlexanderDokuchaev — APPROVED** [review]

**Comment by LeonidBeynenson** [comment]
Updated face detection README-s -- add info on pruning back

Please, note that this PR **should NOT** be merged before the mmdetection PR
https://github.com/openvinotoolkit/mmdetection/pull/149 (also updating documentation only),
since this PR depends on that one.

File: models/image_classification/model_templates/custom-classification/README.md [hunk 11]
```diff
@@ -190,3 +190,29 @@ python export.py \
 
 This produces model `model.xml` and weights `model.bin` in single-precision floating-point format
 (FP32). The obtained model expects **normalized image** in planar BGR format.
+
+
+### 8. Optimization
+
+The models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, only one compression methods is supported for image classification models: int8 quantization.
```
Comment (MaximProshin) [inline models/image_classification/model_templates/custom-classification/README.md:199]: methods => method

**Review by MaximProshin — CHANGES_REQUESTED** [review]

**Comment by MaximProshin** [comment]
Additional general comment: the main OTE page doesn't say anything about the optimization through NNCF. I think there should be a short description that it's supported. I think we also need to clearly explain what the OTE user should do to have NNCF installed together with OTE.

File: models/image_classification/model_templates/custom-classification/README.md (hunk 11 above)
Comment (LeonidBeynenson) [reply models/image_classification/model_templates/custom-classification/README.md:199]: Done

File: models/object_detection/model_templates/face-detection/readme.md [hunk 12]
```diff
@@ -132,3 +132,35 @@ Try both following variants and select the best one:
    * If you would like to start **training** from pre-trained weights use `--load-weights` pararmeter instead of `--resume-from`. Also you can use parameters such as `--epochs`, `--batch-size`, `--gpu-num`, `--base-learning-rate`, otherwise default values will be loaded from `${MODEL_TEMPLATE}`.
 
 As soon as training is completed, it is worth to re-evaluate trained model on test set (see Step 4.b).
+
+
+### 5. Optimize
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply models/object_detection/model_templates/face-detection/readme.md:137]: Done

File: models/object_detection/model_templates/face-detection/readme.md [hunk 13]
```diff
@@ -132,3 +132,35 @@ Try both following variants and select the best one:
...
+
+The models can be optimized -- compressed by [NNCF](https://github.com/openvinotoolkit/nncf) framework.
+
+At the moment, only one compression methods is supported for all face detection models: int8 quantization.
+Also the models face-detection-0200, face-detection-0202, face-detection-0204, and face-detection-0205
+support such NNCF compression method as filter pruning.
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply models/object_detection/model_templates/face-detection/readme.md:143]: Done

[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply models/object_detection/model_templates/face-detection/readme.md:143]: See also PR in mmdetection repo: https://github.com/openvinotoolkit/mmdetection/pull/151

File: models/image_classification/model_templates/custom-classification/README.md [hunk 14]
```diff
@@ -190,3 +190,29 @@ python export.py \
...
+
+Please, note that NNCF framework requires a dataset for compression, since it makes several steps of fine-tuning after
```
[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply models/image_classification/model_templates/custom-classification/README.md:201]: Done

**Comment by LeonidBeynenson** [comment]
>Additional general comment: the main OTE page doesn't say anything about the optimization through NNCF. I think there should be a short description that it's supported. I think we also need to clearly explain what the OTE user should do to have NNCF installed together with OTE.

2 @MaximProshin : Maxim, I added a short description in the main README file of OTE.
Please, note that it is short, since the whole this README file of OTE is very brief.

**Comment by LeonidBeynenson** [comment]
2 @MaximProshin , @Ilya-Krylov : please, note that this PR depends on the mmdetection PR https://github.com/openvinotoolkit/mmdetection/pull/151

**Comment by MaximProshin** [comment]
@LeonidBeynenson , why Instance Segmentation documentation doesn't have the Optimization section? I see that most of the models have the compression config.

File: models/image_classification/model_templates/custom-classification/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (MaximProshin) [reply models/image_classification/model_templates/custom-classification/README.md:210]: I still believe this comment wasn't addressed. The following text is confusing: 
_Please, note that NNCF framework requires a dataset for compression, since it makes several steps of fine-tuning after
compression to restore the quality of the model, so the command line parameters of the script `compress.py` are closer
to the command line parameter of the training script for fine-tuning scenario 4.c stated above, but the number of epochs
required for NNCF compression should not be set by command line parameter, since it is calculated by the script
`compress.py` itself_
Why do we need to say it at all? Can we simply say that the compress script takes the same parameters as the train script?

[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply models/image_classification/model_templates/custom-classification/README.md:210]: Maxim, we could say that, but I'm not sure that this will make the documentation better. 

In my opinion, if a person did not work with compression algorithms, then his/her "default" understanding of compression algorithm is that they do not require training at all, so the dataset is not required. 
(At least this was my understanding before I started work with NNCF/POT)
In this case it will not be clear (for such people) why the compress script's command line parameters likes to the command line parameters of the training script.
It will look like we hide some important details, whereas it is easy to describe them explicitly.

I will make some re-wording of this paragraph to make it shorter, but I still suppose that it is better to keep it.

**Comment by LeonidBeynenson** [comment]
2 @MaximProshin :
>@LeonidBeynenson , why Instance Segmentation documentation doesn't have the Optimization section? I see that most of the models have the compression config.

Added optimization section to the README file for `instance_segmentation/coco-instance-segmentation` problem folder.
Please, note that only two models have NNCF compression enabled in their `template.yaml` files:
`instance-segmentation-0002` and `instance-segmentation-0228`.

The models `instance-segmentation-1039` and `instance-segmentation-1040` have compression config in their folders, but their template files `template.yaml` does not contain valid section `optimizations` -- so, OTE scripts cannot run NNCF compression for them. 
So I propose to change this readme file for these two models only when somebody check if NNCF compression really work for them and (if yes) add optimization section to their `template.yaml` files.

**Comment by MaximProshin** [comment]
> 2 @MaximProshin :
> 
> > @LeonidBeynenson , why Instance Segmentation documentation doesn't have the Optimization section? I see that most of the models have the compression config.
> 
> Added optimization section to the README file for `instance_segmentation/coco-instance-segmentation` problem folder.
> Please, note that only two models have NNCF compression enabled in their `template.yaml` files:
> `instance-segmentation-0002` and `instance-segmentation-0228`.
> 
> The models `instance-segmentation-1039` and `instance-segmentation-1040` have compression config in their folders, but their template files `template.yaml` does not contain valid section `optimizations` -- so, OTE scripts cannot run NNCF compression for them.
> So I propose to change this readme file for these two models only when somebody check if NNCF compression really work for them and (if yes) add optimization section to their `template.yaml` files.

I agree, we should describe it only for those models where the compression works.

File: models/image_classification/model_templates/custom-classification/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (MaximProshin) [reply models/image_classification/model_templates/custom-classification/README.md:210]: I don't propose to remove the paragraph. I propose to make it more concrete and avoid confusing phrases like "the command line parameters of the script compress.py are closer
to the command line parameter of the training script for fine-tuning scenario". Can we say exactly which parameters should be taken by the user from the training/finetuning?

[the remark this answers is not in the document]
Comment (LeonidBeynenson) [reply models/image_classification/model_templates/custom-classification/README.md:210]: Maxim, it is possible to add more information on parameters for compression, but I have to say that at the moment all the other sections (how to start training, finetuning, how to make export, how to evaluate models) do not contain description of the parameters at all -- only examples how to call the scripts.
So, I just made the description for `compress.py` be similar to descriptions of OTE scripts in other sections (or tried to make it a bit more detailed)

Probably I chose not best wording in this section, -- please, correct it.

Unfortunately I'm in vacation from July 26 and will not be able to be online from today.
But if you (or somebody other) propose changes in the documentation, Ilya Krylov will be able to accept them and merge this PR.
